ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -436,109 +440,10 @@ addBlock cfg blk m = Model {
           (currentChain m)
       $ consideredCandidates
 
-    -- See note [Getting the valid blocks] just below
     valid' = foldl
                (Chain.foldChain (\s b -> Set.insert (blockHash b) s))
                (valid m)
```

Comment:
I'm hesitant towards this, because the model could "forget" valid blocks if the blocks are on a chain that isn't connected to the Genesis block anymore. That was the motivation for https://github.com/input-output-hk/ouroboros-network/pull/3651#issue-1158309530.

Assume we use `Set.empty` instead. `Test.O.C.ChainDB.StateMachine` mentions the following about garbage collection:

https://github.com/input-output-hk/ouroboros-consensus/blob/f658b1986cd0035ffad84e04cda52c905cba8b56/ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/ChainDB/StateMachine.hs#L466-L474

Garbage collection on the model is instantaneous, while garbage collection on the SUT is not. The real VolatileDB knows about the validity of all blocks in it, even if those blocks are disconnected.

https://github.com/input-output-hk/ouroboros-consensus/blob/f658b1986cd0035ffad84e04cda52c905cba8b56/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/API.hs#L209-L211

 If we ask the SUT if a disconnected block is valid, and it returns `Just True`, then the model will return `Nothing` (the model doesn't know), because the model only validated connected blocks.

Reply:
I'm now also unsure whether we should have the `Map.member hash (volatileDbBlocks m)` check in `Model.getIsValid`, because it would suffer from the same non-determinism

Reply:
Very good point, I went back to 90b3bdfb5a83e4b55288fff66e9d72035cd29cdb, where I can reproduce https://github.com/input-output-hk/ouroboros-network/issues/3389. Now it turns out that c8a6186feacc14405bdac1386dec38de13427ede (what I originally thought would be enough as the entire fix, ie that was the fifth commit in #284 before I force-pushed it) does *not* fix the bug.

Instead, maintaining the set of valid blocks cumulatively *does* fix the bug, I pushed that know to #284. Note that it does not restrict the model `getIsValid` to just blocks in the VolDB.

> I'm now also unsure whether we should have the `Map.member hash (volatileDbBlocks m)` check in `Model.getIsValid`, because it would suffer from the same non-determinism

So I think we could indeed remove this.